aiohttp/compression_utils.py

```diff
@@ -24,14 +24,95 @@
 MAX_SYNC_CHUNK_SIZE = 1024
 
 
+class ZLibCompressorObjProtocol(Protocol):
+    def compress(self, data: Buffer) -> bytes: ...
+    def flush(self, mode: int = ...) -> bytes: ...
+
+
+class ZLibDecompressorObjProtocol(Protocol):
+    def decompress(self, data: Buffer, max_length: int = ...) -> bytes: ...
+    def flush(self, length: int = ...) -> bytes: ...
+
+    @property
+    def eof(self) -> bool: ...
+
+
+class ZLibBackendProtocol(Protocol):
+    MAX_WBITS: int
+    Z_FULL_FLUSH: int
+    Z_SYNC_FLUSH: int
+    Z_BEST_SPEED: int
+    Z_FINISH: int
+
+    def compressobj(self, *args: Any, **kwargs: Any) -> ZLibCompressorObjProtocol: ...
+    def decompressobj(
+        self, *args: Any, **kwargs: Any
+    ) -> ZLibDecompressorObjProtocol: ...
+
+    def compress(self, data: Buffer, *args: Any, **kwargs: Any) -> bytes: ...
```

Comment:
Shouldn't these match the zlib signatures?
e.g. 
```suggestion
    def compress(self, data: Buffer, /, level: int = ..., wbits: int = ...) -> bytes: ...
```

Reply:
I was on the fence about that - what if zlib changes its signature? I chose to be only as strict as aiohttp uses. Though by that logic I should probably add `*args, **kwargs` to my `ZLibCompressorObjProtocol` and `ZLibDecompressorObjProtocol` functions

Reply:
Only as strict as aiohttp uses makes sense to me, but currently this is incredibly broad and will allow any arguments, which we know are invalid. I don't mind restricting it to the minimal subset needed, but a load of Anys is just avoiding any type checks.

Reply:
I think the way I have it is minimal subset, right? It is saying "I don't care what your library does, as long as it knows how to compress a data `Buffer` and has default values for all other args".

Let's say I take exactly zlib's
```
def compress(self, data: Buffer, /, level: int = ..., wbits: int = ...) -> bytes: ...
```
it'd eclipse any library that implements exactly that plus some magical argument they need to operate.

Likewise, removing the `Any`s and using just what we use
```
def compress9self, data: Buffer)
```
I think would ultimately make it so you can't even use zlib anymore.

Granted, I haven't actually tried the above to be sure, but that is the behavior I expect.

Reply:
Perhaps the right answer is
```
def compress(self, data: Buffer, /, **kwargs: Any) -> bytes: ...
```

Reply:
I'm looking at from the other side, with these annotations aiohttp can call `backend.compress(b"foo", foo=1, bar=7)` and no type error would occur, but it's clearly going to break at runtime.

> Likewise, removing the Anys and using just what we use

My expectation is that a function with signature `(self, data: Buffer, /, level: int = ..., wbits: int = ...)` is compatible with a protocol of `(self, data: Buffer)`, therefore using the minimal requirements, rather than the broadest ones should still type check correctly.

The protocol just says it needs to accept a positional argument. Allowing other optional arguments doesn't break that protocol, it still allows what the protocol is asking for.

Reply:
I see what you're saying. So it's a bit of a chicken and egg problem then. I was using the protocol as a guardrail for what users pass to `set_zlib_backend` whereas you see it as a guardrail for how we use it.

I suppose users can always `cast` if what they want to use a library that doesn't match _exactly_ right?

Reply:
Well, I'm saying that it should work in both directions with my suggestion, because a protocol is the minimum set that must be supported (i.e. aiohttp can't use things not in there, and implementations must support everything in there). So, those implementations should be compatible with the proposed protocol.

However, I don't think mypy is actually validating the modules against that protocol at all, as the zlib module is _not_ compatible with the current protocol. The current protocol says compress() must be able to support data as a keyword, but zlib only supports it as a positional, so it _should_ be failing when you use zlib here.

Reply:
Oh, mypy definitely complained when I tried to pass `asyncio` (chosen arbitrarily) as paraemeter to `set_zlib_backend` so it definitely does some amount of validation.

In any case, I took your suggestions and nothing exploded, if it gets tedious to maintain we always have the ability to change our mind later.